### PR TITLE
Subscribe multi

### LINKED ISSUE #6668: MetaMask provider sends multiple duplicate contract events to subscribers
**Describe the bug**
Multiple events are emitted when using the MetaMask provider and subscribing to a contract event. The number of duplicate events seems to be the product of the number of events subscribed and the number of events emitted by the solidity contract.

## To Reproduce
[Instructions are in this github repository](https://github.com/cds-consensys/mm-contract-event-dupes)

## Expected behavior
As a dapp developer, I want to be notified once per Solidity contract event.

### Single event emitted by Smart Contract
Screen Shot 2019-05-30 at 9 04 45 AM

This screenshot shows side-by-side the result of clicking the `SimpleStorage` button which triggers one event, `SimpleSet` to be fired by the `SimpleStorage` Smart Contract. 
  * Chrome browser with MetaMask extension
      * There are 4 (4*1) events that the listener consumed. There are 4 events that were subscribed to, and the contract emits one event.
  * Chrome incognito sans MetaMask
    * There is only one event consumed, the single event emitted by the smart contract

### Two events emitted by Smart Contract
Screen Shot 2019-05-30 at 9 09 39 AM

This screenshot shows side-by-side the result of clicking the `FimpleStorage` button which triggers two events, `FimpleSet` and `FimpleSet2` to be fired from the `FimpleStorage` Smart Contract ((don't judge, names are hard). 
  * Chrome browser with MetaMask extension
      * There are 8 (4*2) events that the listener consumed. There are 4 events that were subscribed to, and the contract emits 2 event. 
  * Chrome incognito sans MetaMask
    * displays the only two events the Smart Contract emitted

## Video
[Duplicate contract events with MetaMask provider](https://www.youtube.com/watch?v=7KfEtBwCCVc)

## Browser details 
 - OS: OS X v10.14.5
 - Browser: Google Chrome: Version 74.0.3729.169 (Official Build) (64-bit)
 - MetaMask Version: 6.5.3

### PR DESCRIPTION
**Pull request by BelfordZ**
## **Description**

When using the per-dapp network selection feature flag, subscribe and filter requests are not respecting the per-dapp selection. This the main goal of this PR. Additionally, this PR correctly uses the persisted values for the feature flag. It also addresses an issue where changing the feature flag value would require dapps to be refreshed.  

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/6668 https://github.com/MetaMask/MetaMask-planning/issues/1870 https://github.com/MetaMask/MetaMask-planning/issues/1871 https://github.com/MetaMask/MetaMask-planning/issues/1865

## **Manual testing steps**

**note** to test subscribe:
  1. go to https://docs.metamask.io/wallet/reference/eth_subscribe/
  2. open dev tools / console and add event listener: `window.ethereum.on('message', ({data}) => console.log(parseInt(data.result.number, 16)));`
  3. hit execute on eth_subscribe
  4. console should now start showing block updates

  1. Basic Functionality Tests:
    - Verify that subscribing to different Ethereum events (like block headers, logs, or specific event types) works as expected.
    - Test filtering functionalities, ensuring that the filter correctly captures and displays events based on provided criteria.
  3. Network-Specific Tests:
    - Test the subscribe and filter functionalities across different networks (mainnet, Ropsten, Kovan, etc.) to ensure consistent behavior.
    - Check for any network-specific issues or inconsistencies in the functioning of the subscribe and filter features.
  5. Test with and without useRequestQueue Feature Flag:
    - Conduct tests with the useRequestQueue feature flag enabled and verify the functioning of the subscribe and filter features.
    - Repeat the tests with the feature flag disabled to ensure that the functionalities remain stable and consistent.
  6. Repeat the above tests, but in a multi-dapp scenario
  

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

**Comment by BelfordZ** [comment]
@SocketSecurity ignore npm/@metamask/eth-json-rpc-filters@7.0.0

**Review by shanejonas — DISMISSED** [review]

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

**Comment by BelfordZ** [comment]
@shanejonas had to fix rebase on lavamoat files... needs reapprove

**Comment by adonesky1** [comment]
Caught a weird related bug when I took it for a spin

Uploading Screen Recording 2024-01-24 at 10.40.09 AM.mov…

**Comment by adonesky1** [comment]
Caught a weird related bug when I took it for a spin

Uploading Screen Recording 2024-01-24 at 10.40.09 AM.mov…

**Comment by adonesky1** [comment]
Caught a weird bug when I took it for a spin: 

https://github.com/MetaMask/metamask-extension/assets/34557516/6249f58a-20fd-44d3-a02d-94c33ce214e2

**Comment by jiexi** [comment]
I get a different bug where the provider never responds inpage when the domain hasn't visited beforehand with the experimental toggle turned off

https://github.com/MetaMask/metamask-extension/assets/918701/fa3d262e-8b72-4c90-8852-5950da63c1ea


The only way to get it to respond is to visit the new domain with the toggle off, then the domain will work again with the toggle on (including when extension is refreshed)

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

File: test/e2e/tests/state-snapshots/errors-after-init-opt-in-background-state.json [hunk 1]
```diff
@@ -179,8 +179,7 @@
     "selectedAddress": "string"
   },
   "SelectedNetworkController": {
-    "domains": { "metamask": "networkConfigurationId" },
-    "perDomainNetwork": "boolean"
```
Comment (BelfordZ) [inline test/e2e/tests/state-snapshots/errors-after-init-opt-in-background-state.json:183]: I cannot figure out why this goes away now...

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

File: package.json [hunk 2]
```diff
@@ -287,11 +290,11 @@
     "@metamask/post-message-stream": "^7.0.0",
     "@metamask/ppom-validator": "^0.22.0",
     "@metamask/providers": "^14.0.2",
-    "@metamask/queued-request-controller": "^0.3.0",
+    "@metamask/queued-request-controller": "^0.2.0",
```
Comment (adonesky1) [inline package.json:293]: 🤔 why are we downgrading this package?

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

File: app/scripts/lib/middleware/pending.js [hunk 3]
```diff
@@ -2,6 +2,8 @@ import { createAsyncMiddleware } from 'json-rpc-engine';
 import { formatTxMetaForRpcResult } from '../util';
 
 export function createPendingNonceMiddleware({ getPendingNonce }) {
+  // this would not be using the provider for the current domain...
```
Comment (jiexi) [inline app/scripts/lib/middleware/pending.js:5]: This should be okay as the network would be switched to the current domain when this method is called?

File: package.json [hunk 4]
```diff
@@ -326,7 +329,6 @@
     "deep-freeze-strict": "1.1.1",
     "end-of-stream": "^1.4.4",
     "eth-ens-namehash": "^2.0.8",
-    "eth-json-rpc-filters": "^6.0.0",
```
Comment (jiexi) [inline package.json:329]: i think the pattern has generally been to bump deps (and make any necesarry changes to adopt) in a PR by themselves. Just checking if that applies here

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -484,13 +480,26 @@ export default class MetamaskController extends EventEmitter {
...
+      state: initState.SelectedNetworkController,
       messenger: this.controllerMessenger.getRestricted({
         name: 'SelectedNetworkController',
         allowedActions: ['NetworkController:getNetworkClientById'],
         allowedEvents: ['NetworkController:stateChange'],
       }),
     });
 
+    if (
+      !this.selectedNetworkController.getNetworkClientIdForDomain('metamask')
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:492]: Should we pull in and use the `ORIGIN_METAMASK` constant here, using the same variable as the SelectedNetworkController itself. Currently its not exported from the `SelectedNetworkController` so won't block on this but maybe we should export it?

File: app/scripts/lib/middleware/pending.js (hunk 3 above)
Comment (Gudahtt) [reply app/scripts/lib/middleware/pending.js:5]: It can also be called from dapps via `eth_getTransactionCount`. But in that case as well, it should be processed using the provider for the dapp selected network, so I think we're good there too.

**Comment by adonesky1** [comment]
Taking out for a test drive. So far its looking good! Only issue I've seen so far is the `switchEthereumChain` confirmation seems to get swallowed when a dapp calls `switchEthereumChain` to switch to the network the wallet is already on.

**Comment by adonesky1** [comment]
Ooof actually I'm seeing something that doesn't seem very desirable. Any and all domains that I visit are getting added to the `domains` state in the `SelectedNetworkController` regardless of whether I connect to them:

https://github.com/MetaMask/metamask-extension/assets/34557516/dd6b132f-5aeb-48c8-8b4b-b6796a0cc1ed

**Comment by adonesky1** [comment]
> Ooof actually I'm seeing something that doesn't seem very desirable. Any and all domains that I visit are getting added to the domains state in the SelectedNetworkController regardless of whether I connect to them:

@BelfordZ and I have discussed a solution for this which he will push up shortly!

**Comment by adonesky1** [comment]
Found another weeeeirrrrd scenario:

https://github.com/MetaMask/metamask-extension/assets/34557516/a5fa86ec-10f1-42fe-9c9a-3a8bf8e2383f

File: test/e2e/tests/onboarding.spec.js [hunk 6]
```diff
@@ -305,7 +305,10 @@ describe('MetaMask onboarding @no-mmi', function () {
           text: notificationMessage,
         });
         assert.equal(networkNotification, true);
-        await driver.clickElement({ text: 'Dismiss', tag: 'h6' });
+        await driver.clickElement({
```
Comment (Gudahtt) [inline test/e2e/tests/onboarding.spec.js:308]: Hmm, not following why this change was needed

**Comment by Gudahtt** [comment]
>@BelfordZ and I have discussed a solution for this which he will push up shortly!

It would be great if we could get this merged first instead; we already have multiple unrelated bug fixes in this PR already. Ideally we'd be using separate PRs for each one.

**Comment by Gudahtt** [comment]
>going to try and see if I can figure out what's going on here.

Is that on `develop` already as well? I think I recall encountering that last week

**Comment by adonesky1** [comment]
> It would be great if we could get this merged first instead; we already have multiple unrelated bug fixes in this PR already. Ideally we'd be using separate PRs for each one.

yeah this makes sense to me.. This PR has been sitting here for way too long. And definitely makes sense to break it up more going forward. Sorry this has been my fault encouraging scope creep here.

File: test/e2e/tests/onboarding.spec.js (hunk 6 above)
Comment (BelfordZ) [reply test/e2e/tests/onboarding.spec.js:308]: the test didnt really make sense to begin with, since why would you dismiss the dialog that says 'switch to localhost', and then expect it to have switched anyways? 

Im not sure why it was passing to begin with, nor do I really know why it started failing because of this PR.

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Is it possible this wasn't a switchNetworkConfirmation previously?

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Going to look into this.

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Screenshot 2024-02-06 at 12 46 11 PM

so this is what things look like right before `dismiss` is clicked in the test **on develop**

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Screenshot 2024-02-06 at 12 54 37 PM

and this is what I see on this branch... clicking dismiss seems to work fine. It is weird that this switch confirmation modal is coming up when the network is already switched.

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Oh but the ETH balance is not showing up unless you click switch now... even though you appear to already be on localhost 8546

Comment (adonesky1) [reply test/e2e/tests/onboarding.spec.js:308]: Ok so still not sure why this PR broke this test but here is a PR against develop to fix an underlying bug with the add custom network manually flow: https://github.com/MetaMask/metamask-extension/pull/22832

cc @Gudahtt

File: app/scripts/lib/setupSentry.js [hunk 7]
```diff
@@ -225,7 +225,7 @@ export const SENTRY_BACKGROUND_STATE = {
     useTokenDetection: true,
     useRequestQueue: true,
   },
-  SelectedNetworkController: { domains: true, perDomainNetwork: false },
+  SelectedNetworkController: { domains: false, perDomainNetwork: true },
```
Comment (adonesky1) [inline app/scripts/lib/setupSentry.js:228]: We do not want to include `domains` in our sentry reports since it is potentially sensitive and violates our goal to preserve privacy, but we do want to include the `perDomainNetwork` toggle since it is not sensitive and potentially useful for diagnosing bugs.

File: app/scripts/metamask-controller.js [hunk 8]
```diff
@@ -488,13 +484,26 @@ export default class MetamaskController extends EventEmitter {
...
+    ) {
+      this.selectedNetworkController.setNetworkClientIdForMetamask(
+        this.networkController.state.selectedNetworkClientId,
+      );
+    }
+    const networkClient =
+      this.selectedNetworkController.getProviderAndBlockTracker('metamask');
+    this.provider = networkClient.provider;
+    this.blockTracker = networkClient.blockTracker;
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:505]: Does this mean we're now using the clientRegistry for the provider in all of our controllers? Is this desired at this stage?

File: app/scripts/lib/middleware/pending.js (hunk 3 above)
Comment (BelfordZ) [reply app/scripts/lib/middleware/pending.js:5]: it ends up getting (this one)[ 
https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L5211] for `getPendingNonce` - which is not aware of domain.

Comment (BelfordZ) [reply app/scripts/lib/middleware/pending.js:5]: this is kind of getting into cross network accounts though... so yeah not sure how big a problem it really is, thats why I put this comment as a reminder to look into it

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

**Comment by adonesky1** [comment]
Attempting to separate out the `domains` state pollution bug fix + migration into a separate PR here: https://github.com/MetaMask/metamask-extension/pull/22860

File: app/scripts/metamask-controller.js [hunk 9]
```diff
@@ -488,13 +484,32 @@ export default class MetamaskController extends EventEmitter {
...
+        this.networkController.state.selectedNetworkClientId,
+      );
+    }
+
+    let networkClient;
+    if (this.selectedNetworkController.state.perDomainNetwork) {
+      networkClient =
+        this.selectedNetworkController.getProviderAndBlockTracker('metamask');
+    } else {
+      networkClient = this.networkController.getProviderAndBlockTracker();
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:508]: These two network clients should be the same, right? They're both proxied clients representing the globally selected network. Is there any functional impact to which of these branches is taken?

**Comment by adonesky1** [comment]
replaced by https://github.com/MetaMask/metamask-extension/pull/23386
